utils/uri-builder.js

```diff
@@ -1,50 +1,37 @@
 const appRoot = require('app-root-path');
 const config = require('config');
-const decodeUriComponent = require('decode-uri-component');
+const queryString = require('query-string');
 const url = require('url');
 
 const { openapi: { basePath } } = appRoot.require('utils/load-openapi');
 
 const { protocol, hostname } = config.get('server');
 
 /**
- * @summary Self link builder
- * @function
- * @param {string} id resource ID
- * @param {string} resourcePath resource path
- * @returns A self link URL
+ * @summary API base URL
  */
-const selfLink = (id, resourcePath) => url.format({
-  protocol,
-  hostname,
-  pathname: `${basePath}/${resourcePath}/${id}`,
-});
+const apiBaseUrl = url.format({ protocol, hostname, pathname: basePath });
 
 /**
- * @summary Top level query link builder
+ * @summary Resource path link builder
  * @function
- * @param {object} query
+ * @param {string} baseUrl base URL
  * @param {string} resourcePath resource path
- * @returns A decoded url formatted with query parameters in the query object.
+ * @returns A resource path URL
  */
-const querySelfLink = (query, resourcePath) => decodeUriComponent(url.format({
-  protocol,
-  hostname,
-  pathname: `${basePath}/${resourcePath}`,
-  query,
-}));
+const resourcePathLink = (baseUrl, resourcePath) => `${baseUrl}/${resourcePath}`;
 
 /**
- * @summary Paginated link builder
+ * @summary Params link builder
  * @function
- * @param {number} pageNumber Page number of results
- * @param {number} pageSize Number of results to return
- * @param {string} resourcePath resource path
- * @returns A decoded paginated link URL
+ * @param {string} baseUrl base URL
+ * @param {string} params query params
+ * @returns A decoded url formatted with query parameters in the query object
  */
-const paginatedLink = (pageNumber, pageSize, resourcePath) => querySelfLink({
-  'page[number]': pageNumber,
-  'page[size]': pageSize,
-}, resourcePath);
+const paramsLink = (baseUrl, params) => `${baseUrl}?${queryString.stringify(params, { encode: false })}`;
```

Comment:
Is this correct? Shouldn't the parameters be encoded?

Reply:
Yes, this is mainly for the pagination params and it's from the skeleton. By default, the `encode` option is `true` and we will get the following results if we sent this request: ` https://localhost:8080/api/v1/pets?page[size]=50`:

```json
"links": {
    "self": "https://localhost/v1/pets?page%5Bnumber%5D=1&page%5Bsize%5D=50",
    "first": "https://localhost/v1/pets?page%5Bnumber%5D=1&page%5Bsize%5D=50",
    "last": "https://localhost/v1/pets?page%5Bnumber%5D=1&page%5Bsize%5D=50",
    "next": "https://localhost/v1/pets?page%5Bnumber%5D&page%5Bsize%5D=50",
    "prev": "https://localhost/v1/pets?page%5Bnumber%5D&page%5Bsize%5D=50"
},
```

We should encode the request params for sure, but this is not ideal for the return response because of the readability concern. By disabling it, links will be:

```json
"links": {
    "self": "https://localhost/v1/pets?page[number]=1&page[size]=50",
    "first": "https://localhost/v1/pets?page[number]=1&page[size]=50",
    "last": "https://localhost/v1/pets?page[number]=1&page[size]=50",
    "next": "https://localhost/v1/pets?page[number]&page[size]=50",
    "prev": "https://localhost/v1/pets?page[number]&page[size]=50"
},
```